### PR TITLE
[ASNodeController] First implementation of node controller class

### PR DESCRIPTION
**Pull request by hannahmbanana**
Paired programming with @appleguy. Example project - ASDKgram modified to use a node controller:

[ASDKgram-nodeController.zip](https://github.com/facebook/AsyncDisplayKit/files/736714/ASDKgram-nodeController.zip)

### CODE REVIEW

File: AsyncDisplayKit/ASDisplayNode+Subclasses.h [hunk 1]
```diff
@@ -39,7 +39,61 @@ NS_ASSUME_NONNULL_BEGIN
  * variables.
  */
 
-@interface ASDisplayNode (Subclassing)
+@protocol ASInterfaceState <NSObject>
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: Let's name this e.g. `ASInterfaceStateDelegate`. It's OK for something to be its own delegate so this term applies both to nodes and node controllers.

File: AsyncDisplayKit/ASNodeController.mm [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  if (_node == nil) {
+    [self loadNode];
+  }
+  return _node;
+}
+
+-(void)setNode:(ASDisplayNode *)node
+{
+  _node = node;
+  node.interfaceStateDelegate = self;
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/ASNodeController.mm]: We should unhook from our old node `_node.interfaceStateDelegate = nil`.

**Comment by Adlai-Holler** [comment]
@hannahmbanana Could you rebase? Then we'll land!

**Comment by Adlai-Holler** [comment]
Rebased! Here we go!

**Comment by hannahmbanana** [comment]
Opened Issue #2964 to track Adlai's suggestions.

**Pull request merged**